src/internal_rules.jl

```diff
@@ -558,3 +558,247 @@ function EnzymeRules.reverse(
     xs.dval .= xs.dval[back_inds]
     return (nothing,)
 end
+
+function EnzymeRules.forward(
+    ::Const{typeof(cholesky)},
+    RT::Type{<:Union{Const, DuplicatedNoNeed, Duplicated}},
+    A::Union{Const, Duplicated};
+    kwargs...
+)
+    fact = cholesky(A.val; kwargs...)
+    if RT <: Const
+        return fact
+    end
+    # TODO: This will be a problem for sparse matrices as invL and dL are dense
+    invL = inv(fact.L)
+    # TODO: dL is dense even when L was sparse
+    dL = Matrix(fact.L * LowerTriangular(invL * A.dval * invL' * 0.5 * I))
+    # TODO: Stored as Cholesky, although it isn't a Cholesky factorization
+    dfact = Cholesky(dL, 'L', 0)
+    if RT <: DuplicatedNoNeed
+        return dfact
+    else
```

Comment:
Can you merge these two into one function? It'll reduce duplication. See examples above

Reply:
In the examples above they are split (`sort!`). What happens if `RT` is `Duplicated` and input `BatchDuplicated` and vice versa?

Reply:
Bumping this comment for merging. As discussed on call, the same width is used for all types/returns so one cannot have an input with a different batch size (including duplicated (aka 1) vs batch).

Reply:
I don't know how to resolve this. I tried this:

Reply:
Is there a function to access N at runtime?

Reply:
`EnzymeRules.width(config)`

Reply:
oh, not offhand for forward mode (and we should probably create one), but you can make one like:

```
@inline width(::Duplicated) = 1
@inline width(::BatchDuplicated{T, N}) where {T, N} = N
@inline width(::DuplicatedNoNeed) = 1
@inline width(::BatchDuplicatedNoNeed{T, N}) where {T, N} = N
```

Reply:
probably borrow this? https://github.com/EnzymeAD/Enzyme.jl/blob/443f798457da9295598b47cf2fcfe216ca53a2c7/src/Enzyme.jl#L111 and just pass in all the input/return types

Reply:
Though you don't actually need N here, only if RT <: BatchDuplicated/BatchDuplicatedNoNeed, which you can just get the type parameter from